src/Microsoft.VisualStudio.Editor.Razor/DefaultVisualStudioRazorParser.cs

```diff
@@ -0,0 +1,387 @@
+
+        public override ITextBuffer TextBuffer => _documentTracker.TextBuffer;
+
+        // Used in unit tests to ensure we can be notified when idle starts.
+        internal ManualResetEventSlim NotifyForegroundIdleStart { get; set; }
+
+        // Used in unit tests to ensure we can be notified when idle starts.
+        internal ManualResetEventSlim BlockBackgroundIdleWork { get; set; }
+
+        public override async Task ReparseAsync()
+        {
+            // Can be called from any thread
+
+            if (_dispatcher.IsForegroundThread)
+            {
+                ReparseOnForeground(null);
+            }
+            else
+            {
+                await Task.Factory.StartNew(ReparseOnForeground, null, CancellationToken.None, TaskCreationOptions.None, _dispatcher.ForegroundScheduler);
+            }
+        }
+
+        public void Dispose()
+        {
+            _dispatcher.AssertForegroundThread();
+
+            TextBuffer.Changed -= TextBuffer_OnChanged;
+
+            _documentTracker.ContextChanged -= DocumentTracker_ContextChanged;
+            _parser?.Dispose();
+            _parser = null;
+
+            StopIdleTimer();
+        }
+
+        // Internal for testing
+        internal void DocumentTracker_ContextChanged(object sender, EventArgs args)
+        {
+            _dispatcher.AssertForegroundThread();
+
+            if (_parser != null)
+            {
+                _parser.ResultsReady -= OnResultsReady;
+                _parser.Dispose();
+                _parser = null;
+            }
+
+            if (!_documentTracker.IsSupportedProject)
+            {
+                // Tracker is either starting up, tearing down or wrongfully instantiated.
+                // Either way, the tracker can't act on its associated project, neither can we.
+                return;
+            }
+
+            var projectDirectory = Path.GetDirectoryName(_documentTracker.ProjectPath);
+            _templateEngine = _templateEngineFactory.Create(projectDirectory, ConfigureTemplateEngine);
+            _parser = new BackgroundParser(TemplateEngine, FilePath);
+            _parser.ResultsReady += OnResultsReady;
+            _parser.Start();
+
+            // This is temporary until we own the TagHelper resolution system. At that point the parser will push out updates
+            // via DocumentStructureChangedEvents when contexts change. For now, listeners need to know more information about
+            // the parser. In the case that the tracker does not belong to a supported project the editor will tear down its
+            // attachment to the parser when it recognizes the document closing.
+            foreach (var contextChangeListener in _parserContextChangedListeners)
+            {
+                contextChangeListener.OnContextChanged(this);
+            }
+
+            if (_parser.IsIdle)
+            {
+                ReparseAsync().GetAwaiter().GetResult();
```

Comment:
You should not block here. This will wait for the reparse to complete right?

Reply:
Nope, just blocks till the parse request is enqueued.